### ISSUE TITLE
Question: How to migrate a simple Win32 app to WinUI 3.0

### ISSUE THREAD

**Issue by ackh** [microsoft/microsoft-ui-xaml#3812]
I'm currently attempting to figure out how to port my Win32 application to WinUI 3.0. To better understand how basic things are done I'm trying to turn [this](https://docs.microsoft.com/en-us/windows/win32/gdi/capturing-an-image#code-example) Microsoft Win32 sample app into a C++/WinRT based WinUI 3.0 app.

To do so I have installed the Visual Studio WinUI 3.0 Project Templates and have created a new project from the "Blank App, Packaged (WinUI in Desktop)" template. However, there are many things that are unclear to me and I hope some kind soul might help me along with some hints. I stumbled over the following very basic issues:

1. The `wWinMain` function in the mentioned sample sets up the message loop. Where do I need to set the message loop up in WinUI 3.0?
2. Where would I attach the `WndProc` callback function that processes the messages?
3. The sample draws on the window by obtaining an `HDC` using the `GetDC` function. How does this translate to WinUI 3.0? Am I able to draw into a window by simply obtaining the window's `HDC` as well?

**Comment by StephenLPeters** [comment]
@marb2000 Could you weigh in? Do we have Winui3 xaml islands examples?

**Comment by StephenLPeters** [comment]
@Austin-Lamb and @anawishnoff  fyi

**Comment by janos-ijgyarto** [comment]
Seconded. I also have an old Win32 project that I would like to boost using WinUI, and the sample linked by the OP covers most of the use cases involved. The documentation keeps pointing to XAML Islands, but there is no comprehensive guide that describes, from start to finish, how to take a legacy project (i.e without any configuration for WinUI) and setting it up to use WinUI.

**Comment by anawishnoff** [comment]
Most of our team is on holiday break now, but I'm hoping when we return in Jan someone from the engineering side will be able to answer your specific questions, @ackh. Sorry about the delay here.

@janos-ijgyarto - as for documentation for migrating apps to WinUI 3, I know that this is a gap in our documentation. This is something we're hoping and aiming to provide at the WinUI 3.0 release, so stay tuned.

**Comment by ackh** [comment]
Thanks @anawishnoff for your reply. I understand, it's not an urgent matter.

WinUI 3.0 could simplify my life a lot when it comes to Windows app development which is why I'm very interested in understanding how things are done with it. I see no reason for me to wait until it is finally released when I can already start developing with it now. So, I just hope someone can clarify how to port the Windows sample that I pointed to to WinUI 3.0.

**Comment by Austin-Lamb** [comment]
Hi @ackh!  Thanks for your patience as we were all taking some holiday vacations 😊

To answer your specific questions - in a WinUI app you do not need to write your own WinMain function, or your own message pump/WndProc.  The WinUI framework will manage that for you when you create the Application and Window objects that you see in the WinUI project templates.

The GetDC function is from GDI, which is not the rendering system that WinUI uses, so you won't use that function for WinUI apps.  WinUI apps don't have `HDC` objects.  Instead of the traditional tree of `HWND` 'windows' and `HDC` stuff and so on, WinUI apps are built out of a tree of `UIElement` objects.  You can also use the underlying Composition APIs if you want to get to a lower level of the rendering system.

Instead of starting from the sample app you linked to, I would recommend starting with the WinUI 3.0 Project templates that come in the WinUI VSIX.  They'll get you started with the right project file, a markup file for your window, and all the boilerplate code to spin up the WinUI framework and get started.  There are C++/WinRT project templates.

Generally it's recommended that you build your app using XAML markup (.xaml files) as the XAML markup gets parsed into a binary file format called XBF at compile time, which is the most performant way to create a tree of XAML.  You can also create your types in code (ex: new Button() { Content = "Hello World!" } in C# or the equivalent in C++) if you prefer, it just won't be as fast as markup, nor as productive to develop in since you lose access to some features of the Live Visual Tree tools, so stick to markup where you can.

Does that answer your questions?

**Comment by ackh** [comment]
Hey @Austin-Lamb, thanks for getting back to my questions!

The background information you provided is quite helpful but it leaves me wondering whether I could use my Win32 code in a WinUI 3.0 app at all. I was hoping that I could use WinUI 3.0 to combine my existing Win32 code (which makes heavy use of [StretchBlt](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-stretchblt) and [BitBlt](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-bitblt)) with some of my existing C++/WinRT based UWP controls (based on [SwapChainPanel](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.swapchainpanel?view=winrt-19041))

From what I understand it should be straightforward to port the SwapChainPanel based controls to WinUI 3.0. Somewhere in the release notes I saw that SwapChainPanel is now part of WinUI 3.0 so I guess this should work, However, from what you describe I'm currently clueless on how to bring the Win32 code into WinUI 3.0.

I played around with one of the templates a bit and saw that WinUI _seems_ to be offering a `HWND`:

````cpp
winrt::Microsoft::UI::Xaml::Window _window{ nullptr };

...

void App::OnLaunched(LaunchActivatedEventArgs const&)
{
  HWND hwnd;
  _window = make<MainWindow>();
  _window.as<IWindowNative>()->get_WindowHandle(&hwnd);
  _window.Activate();
}
````

Wouldn't that `HWND` allow me to tap into the WinUI 3.0 based windows and connect them to my Win32 code?

If it is not working that way, could I use WinUI 3.0 in another way so that I can build an app with Win32 code and C++/WinRT based controls or is WinUI 3.0 simply not intended for such use cases?

**Comment by Austin-Lamb** [comment]
@ackh - could you describe your scenario more?  Do you want to have, say, a whole window of your app be WinUI 3-based UI while another window is an old-school `HWND` that you use GDI in?  Or do you want to mix the two UI frameworks within a single top-level window?

If you want to mix them, could you explain more about how?  Do you want a small chunk of WinUI content within your larger GDI/HWND-based app?  Or do you want very rich overlapping regions of WinUI and HWNDs?

**Comment by ackh** [comment]
Ideally, I would like to break up the functionality of my Win32 code into individual user interface elements and wrap them in C++/WinRT to provide a clearly defined interface for them via MIDL 3.0. Hypothetically, I would then combine those Win32 based user interface elements together with my SwapChainPanel based user interface elements in a single window, preferably defined using XAML.

XAML would bind bind everything together and connect the individual user interface elements (Win32 and SwapChainPanel based) using dependency properties. However, these Win32 based controls rely on GDI and hence require `HDC`s. But what you described above does not look particularly encouraging if GDI cannot be used to render user interface elements.

My grand hope for WinUI 3.0 is that it becomes the framework that makes this possible. Eventually, it shouldn't matter what rendering technology is used within an application. A window should be able to host user interface elements drawn using GDI or SwapChainPanel based elements that draw using DirectX or whatever else is used to draw (OpenGL, etc.).

It wouldn't even be dramatic if I need to get rid of GDI as long as I can do the same stuff that I'm using GDI for. However, I specifically use good old Win32 tech because other technologies didn't allow me to achieve the same which is taking screenshots as shown in the [sample](https://docs.microsoft.com/en-us/windows/win32/gdi/capturing-an-image#code-example) I mentioned initially.

**Comment by Austin-Lamb** [comment]
Here's how I think about this - a Win32 app is, at the end of the day, a tree of HWNDs.  When WinUI 3 runs inside a Win32 app, it can fill in one of those HWNDs with XAML content.  Right now, WinUI 3 needs to only fill and own an entire top-level window but that is a temporary limitation, eventually we will get XAML Islands working fully which will enable XAML's HWND to be a child HWND of a larger top-level window.

Once that exists, you'd be able to mix and match to some extent - but as of now XAML only has the ability to be hosted within an HWND, it does not have the ability to host an HWND within itself, so XAML would need to be at the 'leaves' of your compositional tree.

We've discussed internally that at some point we should support hosting an HWND inside of XAML, much like WPF has the HwndHost control.  But this is very difficult and expensive to build right, to ensure it interoperates properly with keyboard input, pointer input, accessibility, and so on.  So we have not yet built that, and it's not currently on our near-term roadmap.

So, to sum up, for now you could prototype with WinUI 3 if you have it filling the whole top-level window, enabling you to learn and get more comfortable with C++/WinRT and XAML and so on.  Then, when XAML Islands come online you'd be able to take your GDI-based app and have XAML as a child HWND, you just wouldn't be able to further nest other legacy HWND-based elements within XAML.

You mentioned DirectX as well, and you can do that now with SwapChainPanel as you said.  SwapChainPanel supports DirectX 11 and 12 content.

**Comment by ackh** [comment]
Thanks a lot for you thoughts about this, @Austin-Lamb. This does sound promising to me and makes me worry less about what WinUI 3.0 is actually going to be. I'll play around a bit more with what is already available and look forward to the release of WinUI 3.0.

**Comment by Austin-Lamb** [comment]
Great.  I'm going to close this issue since it seems like your questions have been answered - if you have more specific questions as you go, feel free to open issues on those.
